src/history/hs.c

```diff
@@ -562,11 +562,12 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_PAGE *page, WT_MUL
     uint32_t i;
     uint8_t *p;
     int nentries;
-    bool squashed;
+    bool sformat, squashed;
 
     prev_upd = NULL;
     session = (WT_SESSION_IMPL *)cursor->session;
     insert_cnt = 0;
+    sformat = btree->value_format[0] == 'S';
```

Comment:
This feels fragile, and like the wrong place to find the information to me - is there any alternative?

Reply:
@agorrod Why you think it is the wrong place?

Reply:
@agorrod Can you explain why you think this is the wrong place?

Reply:
I think the decision should be made in the modify code (src/support/modify.c). I'm also nervous that we are making decisions based on the schema of the btree. The modify code via the API assumes (perhaps enforces?) that modifies only ever apply to either simple `WT_ITEM` or `STRING` based schema. The content in the history store can have tables with more complex schemas, so checking just the first character of the schema for 'S' feels fragile. i.e: what would happen in cases where a table has a schema of `SSqqS`? Should it actually account for three extra offset oddities in the modify code? Are we certain that we aren't every trying to calculate a modify for that case?
I'm also uncertain because looking at the schema of the btree means that we are aware of the data being stored, and we've run into other problems with the history store when assuming relationships between keys for a btree (i.e: custom collator ordering means keys can be in a different order in the history store than they appear in the data file).

Reply:
@agorrod Here the btree is the data btree not the hs betree. Also we are doing the same thing for modify in update chain:
```
int
__wt_modify_apply(WT_CURSOR *cursor, const void *modify)
{
    WT_SESSION_IMPL *session;
    bool sformat;

    session = (WT_SESSION_IMPL *)cursor->session;
    sformat = cursor->value_format[0] == 'S';

    return (__wt_modify_apply_item(session, &cursor->value, modify, sformat));
}
```

Reply:
I agree it is wrong if we check this on history store schema but I think we never do that.

Reply:
@quchenhao I understand which btree is being looked at.
As far as I can see there is nothing tying those things together. At the moment we decide to use reverse modifies when we encounter a modify, but the intent was to *always* use reverse modifies with the history store.
If we make that change to use modify more often in the history store after using your fix here, it will appear to work, but there will be subtle off-by-one bugs.
That's why my comment - I believe this fixes the scenario for today, but the check is in the wrong place because it's makes it more likely we are introducing subtle bugs in the future.

Reply:
I moved the check to modify.c but I don't think currently we can get rid of the assumption that there is one field in the data file schema.